File: algorithms/0374.guess-number-higher-or-lower/solution.rs
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * unsafe fn guess(num: i32) -> i32 {}
 */

impl Solution {
    unsafe fn guessNumber(n: i32) -> i32 {
        let (mut left, mut right) = (0, n);
        let mut mid: i32;

        loop {
            mid = left + (right - left) / 2;
            match guess(mid) {
                -1 => { right = mid - 1; },
                1 => { left  = mid + 1; },
                _ => { return mid; },
            }
        }
    }
}

File: algorithms/0374.guess-number-higher-or-lower/solution.cpp
/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
  public:
    int guessNumber(int n) {
        int left = 0;
        int right = n;

        int res;
        while (left <= right) {
            int mid = left + (right - left) / 2;

            int pick = guess(mid);
            if (pick == -1) {
                right = mid - 1;
            } else if (pick == 1) {
                left = mid + 1;
            } else {
                res = mid;
                break;
            }
        }

        return res;
    }
};
